### PR TITLE
node: watch_log_for: reduce polling sleep time

### PR DESCRIPTION
**Pull request by bhalevy**
Currently, when watch_log_for reaches end-of-file
it sleeps for 10 milliseconds (0.01 seconds).
This proved to be too much as the caller sometimes needs to quickly react to certain log messages, and missing the event in 10 milliseconds will be too late, as seen in https://github.com/scylladb/scylla-dtest/pull/4393#issuecomment-2148934115

This change reduces the sleep time to a single millisecond and also changes the timeout detection to using a proper time-based deadline, rather than counting the number of times we sleep (I'm really not sure why it was originally done that way).

### CODE REVIEW

**Review by yarongilor — APPROVED** [review]
LGTM. i guess we'll have to give a try running some tests to see no surprises following more excessive cycles.

File: ccmlib/node.py
Comment (fruch) [reply ccmlib/node.py:489]: if one insist on changing the default behavier, we should be running all of the tests

Comment (bhalevy) [reply ccmlib/node.py:489]: yes, though a millisecond is quite a long time nowadays :)

**Review by tchaikov — APPROVED** [review]
lgtm

File: ccmlib/node.py
Comment (tchaikov) [reply ccmlib/node.py:489]: agreed. also, i have a python wrapper for inotify. will contribute to ccm once this change lands.

Comment (tchaikov) [reply ccmlib/node.py:489]: @fruch hi Israel, with my change (a future ware, i knew.. but it is sitting in my local repo, i swear.), Benny's change should be fine even if you still have this concern.

**Comment by bhalevy** [comment]
In v2 (0e31589):
- added a `polling_interval` parameter using the existing default (0.01 seconds)
  - can be provided by specific callers that need higher frequency polling.

File: ccmlib/node.py
Comment (bhalevy) [reply ccmlib/node.py]: > agreed. also, i have a python wrapper for inotify. will contribute to ccm once this change lands.

there's also https://pypi.org/project/inotify/

but I think that using it won't be straightforward.
we'll need to install and start the event handler that will probably need to signal some condition variable
and the foreground thread, insead of sleep unconditionally will wait on the condition variable, but still with reasonable timeout to deal with missing events as inotify is not guaranteed to be perfect or race free...

Comment (tchaikov) [reply ccmlib/node.py]: created #578

**Comment by tchaikov** [comment]
hi @fruch and @yarongilor if no objections, i am going to merge this change. as it does not change the existing behavior.

**Comment by yarongilor** [comment]
> > LGTM. i guess we'll have to give a try running some tests to see no surprises following more excessive cycles.
> 
> In v2 the existing behavior shouldn't change (though I don't think polling every millisecond would make any change in tests, but you never know...) This will require passing polling_interval=0.001 in `test_mv_resurrected_rows_after_decommission_interrupt`

perhaps a format of `polling_interval=1ms` or `polling_interval_ms=1` would be better

**Comment by bhalevy** [comment]
> > Currently, when watch_log_for reaches end-of-file it sleeps for 10 milliseconds (0.01 seconds). This proved to be too much as the caller sometimes needs to quickly react to certain log messages, and missing the event in 10 milliseconds will be too late, as seen in [scylladb/scylla-dtest#4393 (comment)](https://github.com/scylladb/scylla-dtest/pull/4393#issuecomment-2148934115)
> > This change reduces the sleep time to a single millisecond and also changes the timeout detection to using a proper time-based deadline, rather than counting the number of times we sleep (I'm really not sure why it was originally done that way).
> 
> How can a test that is so sensitive to timing be stable?

I've just commented there, that the proper to test that would be using error injection to trigger the node restart at the right spot.

That said, we have other tests that wait for node operations to start streaming, based on log messages.
They are less sensitive to timing because they typically wait on a system keyspace, which happens early in the streaming process, or wait for the user keyspace to **start** streaking, not finish it, though in this case 10 ms delay is also tight, and 1ms polling would help prevent flakiness there as well.

**Comment by bhalevy** [comment]
ping, can we get this merged please?

**Comment by tchaikov** [comment]
> > > LGTM. i guess we'll have to give a try running some tests to see no surprises following more excessive cycles.
> > 
> > 
> > In v2 the existing behavior shouldn't change (though I don't think polling every millisecond would make any change in tests, but you never know...) This will require passing polling_interval=0.001 in `test_mv_resurrected_rows_after_decommission_interrupt`
> 
> perhaps a format of `polling_interval=1ms` or `polling_interval_ms=1` would be better

`polling_interval`'s unit is second, which is consistent with that of `timeout`. so i think it's better in the sense of consistency.

i am merging this change. as no objections were raised.

**Comment by tchaikov** [comment]
@scylladb/qa-maintainers hello maintainers, could you help queue this change?

**Review by roydahan — APPROVED** [review]

**Pull request merged**
